### PR TITLE
fix(core): Fix various bugs that left locks stuck.

### PR DESCRIPTION
**Pull request by broneill**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** 
Some unit tests fail because they leave locks stuck, and this also has been observed in the running service itself.

**New behavior :**
Unit tests perform more validation to help narrow down root cause, and a bunch of fixes have been made.

### CODE REVIEW

File: query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala [hunk 1]
```diff
@@ -142,12 +142,29 @@ final case class InstantVectorFunctionMapper(function: InstantFunctionId,
   }
 }
 
+private abstract class FuncIterator(rows: Iterator[RowReader])  extends Iterator[RowReader] {
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala:145]: Have you evaluated impact if you use this abstract class in the RangeVector trait ?

Comment (broneill) [reply query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala:145]: no

Comment (broneill) [reply query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala:145]: What issues do you expect?

Comment (vishramachandran) [reply query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala:145]: I think it is worth a shot. There may be much good out of this. Right now developers can forget to inherit from this interface.

Comment (broneill) [reply query/src/main/scala/filodb/query/exec/RangeVectorTransformer.scala:145]: I'm just going for the simplest changes for now, avoiding any possible destabilization.

**Comment by broneill** [comment]
I created a more generic wrapper class that can be reused, but the "lack of proper failure handling" is so pervasive, there's no quick fix and it's way out of scope for this PR.

**Pull request closed without merging**